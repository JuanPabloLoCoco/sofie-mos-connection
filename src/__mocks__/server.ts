import { EventEmitter } from 'events'

import { Server } from 'net'
import { SocketMock } from './socket'
// import { Writable } from 'stream'

const iconv = require('iconv-lite')
iconv.encodingExists('utf16-be')

// Mock the Server class in 'net':
let setTimeoutOrg = setTimeout

let instances: Array<IServerMock> = []

export class ServerMock extends EventEmitter implements Server {
	// Mock implementation:

	maxConnections: number
	connections: number
	listening: boolean
	listenToPort: number

	private _responses: Array<(data: any) => string | Buffer > = []

	constructor () {
		super()

		// @ts-ignore this is comparable with ISocketMock
		instances.push(this)

		this.listen = jest.fn(this.listen)
		this.close = jest.fn(this.close)
		this.address = jest.fn(this.address)
		this.getConnections = jest.fn(this.getConnections)
		this.ref = jest.fn(this.ref)
		this.unref = jest.fn(this.unref)
		// this.addListener = jest.fn(this.addListener)
		// this.emit = jest.fn(this.emit)
		// this.on = jest.fn(this.on)
		// this.once = jest.fn(this.once)
		// this.prependListener = jest.fn(this.prependListener)
		// this.prependOnceListener = jest.fn(this.prependOnceListener)

	}
	static mockClear () {
		instances.splice(0, 9999)
	}
	static get instances () {
		return instances
	}

	listen (port: any) {

		this.listenToPort = port
		setTimeoutOrg(() => {
			this.listening = true
			this.emit('listening', {})
		},1)
		return this
	}
	close (callback?: Function) {
		if (callback) callback()
		setTimeoutOrg(() => {
			this.emit('close')
		},1)
		return this
	}
	address () { return { port: 0, family: 'string', address: 'string' } }
	getConnections (cb: (error: Error | null, count: number) => void) { cb(null, 0) }
	ref () { return this }
	unref () { return this }
	// addListener (event: string, listener: (...args: any[]) => void) { return this }
	// emit (event: string | symbol, ...args: any[]) { return true }
	// on (event: string, listener: (...args: any[]) => void) { return this }
	// once (event: string, listener: (...args: any[]) => void) { return this }
	// prependListener (event: string, listener: (...args: any[]) => void) { return this }
	// prependOnceListener (event: string, listener: (...args: any[]) => void) { return this }
	// ------------------------------------------------------------------------
	// Mock methods:
	mockNewConnection () {
		// "Someone has connected"
		let socket = new SocketMock()
		socket.on('data', (d) => {
			this.emit('data', d)
		})

		this.emit('connection', socket)

		return socket
	}

	mockSentMessage (data: any, encoding: any) {
		encoding = encoding
		if (this._responses.length) {
			// send reply:

			let cb: any = this._responses.shift()
			let msg

			setTimeoutOrg(() => {

				if (typeof cb === 'string') {
					msg = cb
				} else {
					msg = cb(data)
				}

				this.mockReceiveMessage(msg)
			},1)
		}
	}
	mockReceiveMessage (msg: string | Buffer) {
		this.emit('data', msg)
	}
	mockAddReply (cb: (data: any) => string | Buffer) {
		this._responses.push(cb)
	}
	mockClear () {
		this._responses.splice(0, 9999)
	}
	decode (data: Buffer): string {
		return iconv.decode(data, 'utf16-be')
	}
	encode (str: string) {
		return iconv.encode(str, 'utf16-be')
	}
}

export type IServerMock = jest.Mocked<ServerMock>
