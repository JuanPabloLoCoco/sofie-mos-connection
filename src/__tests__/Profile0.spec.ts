import { clearMocks,
	delay,
	doBeforeAll,
	fakeIncomingMessage,
	getMosConnection,
	getMosDevice,
	setupMocks
} from './lib'
import {
	MosConnection,
	MosDevice,
	IMOSObject,
	IMOSListMachInfo
} from '..'
import { SocketMock } from '../__mocks__/socket'
import { xmlData, xmlApiData } from '../__mocks__/testData'

// @ts-ignore imports are unused
import { Socket } from 'net'

beforeAll(() => {
	setupMocks()
})
beforeEach(() => {
	clearMocks()
})
describe('Profile 0', () => {
	let mosDevice: MosDevice
	let mosConnection: MosConnection

	let socketMockLower: SocketMock
	let socketMockUpper: SocketMock
	let socketMockQuery: SocketMock

	let serverSocketMockLower: SocketMock
	let serverSocketMockUpper: SocketMock
	let serverSocketMockQuery: SocketMock

	let onRequestMachineInfo: jest.Mock<any, any>
	let onRequestMOSObject: jest.Mock<any, any>
	let onRequestAllMOSObjects: jest.Mock<any, any>

	beforeAll(async () => {

		mosConnection = await getMosConnection({
			'0': true,
			'1': true // Must support at least one other profile
		}, true)
		mosDevice = await getMosDevice(mosConnection)

		// Profile 0:
		onRequestMachineInfo = jest.fn(() => {
			return Promise.resolve(xmlApiData.machineInfo)
		})
		mosDevice.onRequestMachineInfo((): Promise<IMOSListMachInfo> => {
			return onRequestMachineInfo()
		})
		// Profile 1:
		onRequestMOSObject = jest.fn(() => {
			return Promise.resolve(xmlApiData.mosObj)
		})
		onRequestAllMOSObjects = jest.fn(() => {
			return Promise.resolve([
				xmlApiData.mosObj,
				xmlApiData.mosObj2
			])
		})
		mosDevice.onRequestMOSObject((objId: string): Promise<IMOSObject | null> => {
			return onRequestMOSObject(objId)
		})
		mosDevice.onRequestAllMOSObjects((): Promise<Array<IMOSObject>> => {
			return onRequestAllMOSObjects()
		})
		let b = doBeforeAll()
		socketMockLower = b.socketMockLower
		socketMockUpper = b.socketMockUpper
		socketMockQuery = b.socketMockQuery
		serverSocketMockLower = b.serverSocketMockLower
		serverSocketMockUpper = b.serverSocketMockUpper
		serverSocketMockQuery = b.serverSocketMockQuery
	})
	afterAll(async () => {
		await mosDevice.dispose()
		await mosConnection.dispose()
	})
	beforeEach(() => {
		onRequestMOSObject.mockClear()
		onRequestAllMOSObjects.mockClear()

		serverSocketMockLower.mockClear()
		serverSocketMockUpper.mockClear()
		if (serverSocketMockQuery) serverSocketMockQuery.mockClear()
		socketMockLower.mockClear()
		socketMockUpper.mockClear()
		if (socketMockQuery) socketMockQuery.mockClear()
	})
	test('init', async () => {
		expect(mosDevice).toBeTruthy()
		expect(socketMockLower).toBeTruthy()
		expect(socketMockUpper).toBeTruthy()
		expect(serverSocketMockLower).toBeTruthy()
	})
	test('heartbeat from other party', async () => {

		expect(serverSocketMockLower).toBeTruthy()

		serverSocketMockLower.setReplyToHeartBeat(false)

		let serverReply: jest.Mock<any, any> = jest.fn(() => false)
		serverSocketMockLower.mockAddReply(serverReply)
		// Fake incoming message on socket:
		let sendMessageId = await fakeIncomingMessage(serverSocketMockLower, xmlData.heartbeat)
		await delay(10) // to allow for async timers & events to triggered

		expect(serverReply).toHaveBeenCalledTimes(1)

		let msg = serverSocketMockLower.decode(serverReply.mock.calls[0][0])

		expect(msg).toMatch(/<heartbeat/)
		expect(msg).toMatch('<messageID>' + sendMessageId)
	})

	test('unknown party connects', async () => {

		// let unknownServerSocketMockLower = serverMockLower.mockNewConnection()

		expect(serverSocketMockLower).toBeTruthy()
		serverSocketMockLower.setReplyToHeartBeat(false)
		let serverReply: jest.Mock<any, any> = jest.fn(() => false)
		serverSocketMockLower.mockAddReply(serverReply)

		// Fake incoming message on socket:
		let sendMessageId = await fakeIncomingMessage(
			serverSocketMockLower,
			xmlData.heartbeat,
			'ourUnknownMosId'
		)
		await delay(10) // to allow for async timers & events to triggered

		expect(serverReply).toHaveBeenCalledTimes(1)
		let msg = serverSocketMockLower.decode(serverReply.mock.calls[0][0])
		expect(msg).toMatch(/<mosAck>/)
		expect(msg).toMatch('<messageID>' + sendMessageId)
		expect(msg).toMatch('<status>NACK')

		serverReply.mockClear()
		serverSocketMockLower.mockAddReply(serverReply)

		// Fake incoming message on socket:
		sendMessageId = await fakeIncomingMessage(
			serverSocketMockLower,
			xmlData.heartbeat,
			undefined,
			'theirUnknownMosId'
		)
		await delay(10) // to allow for async timers & events to triggered

		expect(serverReply).toHaveBeenCalledTimes(1)
		msg = serverSocketMockLower.decode(serverReply.mock.calls[0][0])
		expect(msg).toMatch(/<mosAck>/)
		expect(msg).toMatch('<messageID>' + sendMessageId)
		expect(msg).toMatch('<status>NACK')

	})
	// TODO: reqMachInfo
	// TODO: listMachInfo
})
