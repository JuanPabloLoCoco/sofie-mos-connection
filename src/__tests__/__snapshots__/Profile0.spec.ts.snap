// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile 0 requestMachineInfo 1`] = `
"<mos>
  <ncsID>their.mos.id</ncsID>
  <mosID>our.mos.id</mosID>
  <messageID>xx</messageID>
  <reqMachInfo/>
</mos>"
`;

exports[`Profile 0 requestMachineInfo 2`] = `
Object {
  "DOM": undefined,
  "ID": MosString128 {
    "_str": "airchache.newscenter.com",
    "strict": true,
  },
  "SN": MosString128 {
    "_str": "927748927",
    "strict": true,
  },
  "defaultActiveX": undefined,
  "hwRev": MosString128 {
    "_str": "",
    "strict": true,
  },
  "manufacturer": MosString128 {
    "_str": "RadioVision, Ltd.",
    "strict": true,
  },
  "model": MosString128 {
    "_str": "TCS6000",
    "strict": true,
  },
  "mosExternalMetaData": undefined,
  "mosRev": MosString128 {
    "_str": "2.8.2",
    "strict": true,
  },
  "opTime": MosTime {
    "_time": 2009-03-01T23:55:10.000Z,
    "_timezone": "Z",
    "_timezoneDeclaration": "",
    "_timezoneZuluIndicator": "Z",
  },
  "supportedProfiles": Object {
    "deviceType": "NCS",
    "profile0": true,
    "profile1": true,
    "profile2": true,
    "profile3": true,
    "profile4": true,
    "profile5": true,
    "profile6": true,
    "profile7": true,
  },
  "swRev": MosString128 {
    "_str": "2.1.0.37",
    "strict": true,
  },
  "time": MosTime {
    "_time": 2009-04-11T17:20:42.000Z,
    "_timezone": "Z",
    "_timezoneDeclaration": "",
    "_timezoneZuluIndicator": "Z",
  },
}
`;
